// payhero-devkit/packages/client.ts
import axios, { AxiosInstance, AxiosError } from "axios";
import {
  PayHeroConfig,
  StkPushRequest,
  StkPushResponse,
  TransactionStatusRespone,
  PollingOptions,
  WalletBalanceResponse,
  WalletBalanceRequest,
  PaymentsWalletBalanceResponse,
  ServiceWalletTopUpRequest,
  WalletWithdrawRequest,
  WalletWithdrawResponse
} from "./types";

const base_url = "https://backend.payhero.co.ke/api/v2/payments";
export class PayHeroClient {
  private client: AxiosInstance;

  constructor(config: PayHeroConfig) {
    this.client = axios.create({
      baseURL: base_url,
      headers: {
        Authorization: config.authToken,
        "Content-Type": "Application/json",
      },
    });
  }

  async stkPush(data: StkPushRequest): Promise<StkPushResponse> {
      const required = [
      "phone_number",
      "amount",
      "provider",
      "channel_id",
      "external_reference",
    ];

    for (const fields of required) {
      if (!data[fields as keyof StkPushRequest]) {
        throw new Error(`Missing required field: ${fields}`);
      }
    }
    const formatPhoneNumber = (phoneNumber: number | string): string => {
      const number = String(phoneNumber).replace(/\D/g, "");

      if (number.startsWith("254") && number.length === 12) return number;
      if (number.startsWith("7") && number.length === 9) return `254${number}`;
      if (number.startsWith("0") && number.length === 10)
        return `254${number.slice(1)}`;
      throw new Error("Invalid phone number format");
    };

    const formatedPhoneNumber = formatPhoneNumber(data.phone_number);

    if (data.provider === "sasapay" && data.channel_id && !data.network_code) {
      throw new Error("network_code is required fro sasapay with channel_id");
    }

    const payload: Record<string, any> = {
      phone_number: formatedPhoneNumber,
      amount: data.amount,
      provider: data.provider,
      channel_id: data.channel_id,
      external_reference: data.external_reference,
    };

    if (data.customer_name) {
      payload.customer_name = data.customer_name;
    }

    if (data.credential_id) {
      payload.credential_id = data.credential_id;
    }

    if (data.network_code) {
      payload.network_code = data.network_code;
    }

    const response = await this.client.post(base_url, payload);

    return response.data as StkPushResponse;
  }

  async transactionStatus(
    reference: string
  ): Promise<TransactionStatusRespone> {
    const response = await this.client.get(
      `https://backend.payhero.co.ke/api/v2/transaction-status?reference=${reference}`
    );

    return response.data as TransactionStatusRespone;
  }

  async pollTransactionStatus(
    reference: string,
    options: PollingOptions = {}
  ): Promise<TransactionStatusRespone> {
    const { interval = 5000, maxAttenpt = 12, timeout = 60000 } = options;

    const startTime = Date.now();

    for (let attempt = 1; attempt <= maxAttenpt; attempt++) {
      if (Date.now() - startTime > timeout) {
        throw new Error(`Polling timeout after ${timeout}ms`);
      }

      try {
        const status = await this.transactionStatus(reference);

        if (status.status === "SUCCESS" || status.status === "FAILED") {
          return status;
        }

        await new Promise((resolve) => setTimeout(resolve, interval));
      } catch (err) {
        if (attempt === maxAttenpt) throw err;

        await new Promise((resolve) => setTimeout(resolve, interval / 2));
      }
    }

    throw new Error(`Polling failed after ${maxAttenpt} attempts`);
  }

  async servicewalletBalance(
    wallet_type?: WalletBalanceRequest
  ): Promise<WalletBalanceResponse> {
    try {
      const response = await this.client.get(
        `https://backend.payhero.co.ke/api/v2/wallets?wallet_type=${wallet_type}`
      );

      return response.data as WalletBalanceResponse;
    } catch (error) {
      throw new Error(
        `Error fetching Service wallet balance: ${(error as AxiosError).message
        }`
      );
    }
  }

  async paymentWalletBalance(
    walletChannelID: number
  ): Promise<PaymentsWalletBalanceResponse> {
    if (!walletChannelID || typeof walletChannelID !== "number") {
      throw new Error("WalletChannelID must be a valid number");
    }

    try {
      const response = await this.client.get(
        `https://backend.payhero.co.ke/api/v2/payment_channels/${walletChannelID}`
      );

      return response.data as PaymentsWalletBalanceResponse;
    } catch (error) {
      throw new Error(
        `Failed to fetch Payment wallet balance: ${(error as AxiosError).message
        }`
      );
    }
  }

  async serviceWalletTopup(
    data: ServiceWalletTopUpRequest
  ): Promise<StkPushResponse> {
    const required = ["amount", "phone_number"];

    for (const fields of required) {
      if (!data[fields as keyof ServiceWalletTopUpRequest]) {
        throw new Error(`Missing required fields ${fields}`);
      }
    }
    const formatPhoneNumber = (phoneNumber: number | string): string => {
      const number = String(phoneNumber).replace(/\D/g, "");

      if (number.startsWith("254") && number.length === 12) return number;
      if (number.startsWith("7") && number.length === 9) return `254${number}`;
      if (number.startsWith("0") && number.length === 10)
        return `254${number.slice(1)}`;
      throw new Error("Invalid phone number format");
    };

    const formatedPhoneNumber = formatPhoneNumber(data.phone_number);

    const payload: Record<string, any> = {
      phone_number: formatedPhoneNumber,
      amount: data.amount,
    };
    try {
      const response = await this.client.post(
        "https://backend.payhero.co.ke/api/v2/topup",
        payload
      );

      return response.data as StkPushResponse;
    } catch (error) {
      throw new Error(
        `Error topping up serveice wallet: ${(error as AxiosError).message}`
      );
    }
  }

  async walletWithdraw(data: WalletWithdrawRequest): Promise<WalletWithdrawResponse> {

    const required = [
      "amount",
      "phone_number",
      "network_code",
      "channel",
      "channel_id"
    ]

    for (const fields of required) {
      if(!data[fields as keyof WalletWithdrawRequest]) {
        throw new Error(`Missing required fields ${fields}`)
      }
    }

    const formatPhoneNumber = (phoneNumber: number | string): string => {
      const number = String(phoneNumber).replace(/\D/g, "");

      if (number.startsWith("254") && number.length === 12) return number;
      if (number.startsWith("7") && number.length === 9) return `254${number}`;
      if (number.startsWith("0") && number.length === 10)
        return `254${number.slice(1)}`;
      throw new Error("Invalid phone number format");
    };

    const formatedPhoneNumber = formatPhoneNumber(data.phone_number)

    const payload = {
      amount: data.amount,
      phone_number: formatedPhoneNumber,
      network_code: data.network_code,
      channel: data.channel,
      channel_id: data.channel_id,
      ...(data.external_reference && { external_refernce: data.external_reference}),
      ...(data.callback_url && { callback_url: data.callback_url}),
      ...(data.payment_service && { payment_serveice: data.payment_service})
    }

    try {
      const response = await this.client.post("https://backend.payhero.co.ke/api/v2/withdraw", payload)

      return response.data as WalletWithdrawResponse
    } catch (error) {
      throw new Error(`Error processing withdrawal ${(error as AxiosError)}`)
    }
  }
}
