// payhero-devkit/packages/types.ts
export interface PayHeroConfig {
    authToken: string;
}

export interface StkPushRequest {
    phone_number: string;
    amount: number;
    channel_id: number;
    provider: 'm-pesa' | 'sasapay';
    external_reference: string;
    customer_name?: string;
    callback_url?: string;
    credential_id?: string;
    network_code?: string;
}

export interface StkPushResponse {
    success: boolean;
    status: "QUEUED";
    reference: string;
    CheckoutReferenceID: string;
    external_reference: string;
}

export interface TransactionStatusRespone {
    success: boolean;
    status: string;
    reference: string;
    provider: string;
    merchant?: string;
    payment_reference?: string;
    third_party_refernce?: string;
    CheckoutRequestID?: string;
    provider_refernce?: string;
    transaction_date?: string
}

export interface PollingOptions {
    interval?: number;
    maxAttenpt?: number;
    timeout?: number;
}

export interface WalletBalanceRequest {
    wallet_type?: string | "service_wallet"
}

export interface WalletBalanceResponse {
    id: number;
    account_id: number;
    wallet_type: string | "service_wallet";
    currency: string;
    available_balance: number;
    account_number?: string;
    beneficiary?: string;
    wallet_status?: string;
    wallet_id?: number;
    created_at: Date;
    updated_at: Date;
}

// export interface PaymentsWalletRequest {
//     walletChannelID: number;
// }

export interface PaymentsWalletBalanceResponse {
    id: number;
    transaction_type: string;
    channel_type: string;
    account_id: number;
    short_code: string;
    account_Number: string;
    description: string,
    is_active: boolean;
    balance_plain: {
        last_balance: number;
        balance: number;
        last_reference: string;
    }
    created_at: Date;
    updated_at: Date
}

export interface ServiceWalletTopUpRequest {
    amount: number;
    phone_number: string;
}

export interface WalletWithdrawRequest {
    external_reference?: string;
    amount: number;
    phone_number: string;
    network_code: string;
    callback_url?: string;
    channel: string;
    channel_id: number;
    payment_service?: string;
}

export interface WalletWithdrawResponse {
    status: string | "QUEUED";
    merchant_reference: string;
    checkout_request_id: string;
    response_code: string;
    conversation_id: string;
}
